File: coding/karat-log-access/sol.java
import java.util.*;

public class LogAnalyzer {

    public static Map<String, List<Integer>> getSortedAccessTimes(List<List<String>> logs) {
        Map<String, List<Integer>> accessTimes = new HashMap<>();

        for (List<String> log : logs) {
            String user = log.get(1);
            int num = Integer.parseInt(log.get(0));

            if (accessTimes.containsKey(user)) {
                List<Integer> times = accessTimes.get(user);
                if (times.get(0) > num) {
                    times.set(0, num);
                } else if (times.get(1) < num) {
                    times.set(1, num);
                }
            } else {
                accessTimes.put(user, new ArrayList<>(Arrays.asList(num, num)));
            }
        }
        return accessTimes;
    }

    public static List<String> getMostAccessedResource(List<List<String>> logs) {
        Map<String, List<Integer>> accessTimes = new HashMap<>();
        int max = 0;
        String resourceMax = "";

        for (List<String> log : logs) {
            String resource = log.get(2);
            int num = Integer.parseInt(log.get(0));

            accessTimes.putIfAbsent(resource, new ArrayList<>());
            List<Integer> times = accessTimes.get(resource);

            while (!times.isEmpty() && times.get(0) < num - 300) {
                times.remove(0);
            }
            times.add(num);

            if (times.size() > max) {
                max = times.size();
                resourceMax = resource;
            }
        }

        return Arrays.asList(resourceMax, String.valueOf(max));
    }

    public static void printSolution(Map<String, List<Integer>> accessTimes) {
        for (Map.Entry<String, List<Integer>> entry : accessTimes.entrySet()) {
            System.out.println(entry.getKey() + " " + entry.getValue().get(0) + "," + entry.getValue().get(1));
        }
    }

    public static void main(String[] args) {
        List<List<String>> logs1 = Arrays.asList(
            Arrays.asList("2", "user_1", "resource_1"),
            Arrays.asList("3234", "user_2", "resource_2"),
            Arrays.asList("456", "user_1", "resource_3"),
            Arrays.asList("200", "user_6", "resource_5"),
            Arrays.asList("215", "user_6", "resource_4"),
            Arrays.asList("43553", "user_2", "resource_3"),
            Arrays.asList("63346", "user_3", "resource_3"),
            Arrays.asList("35445", "user_22", "resource_1"),
            Arrays.asList("5345", "user_5", "resource_3"),
            Arrays.asList("2", "user_6", "resource_1"),
            Arrays.asList("100", "user_6", "resource_6"),
            Arrays.asList("400", "user_7", "resource_2"),
            Arrays.asList("100", "user_8", "resource_6"),
            Arrays.asList("5534", "user_1", "resource_3")
        );

        List<List<String>> logs2 = Arrays.asList(
            Arrays.asList("300", "user_1", "resource_3"),
            Arrays.asList("599", "user_1", "resource_3"),
            Arrays.asList("900", "user_1", "resource_3"),
            Arrays.asList("1199", "user_1", "resource_3"),
            Arrays.asList("1200", "user_1", "resource_3"),
            Arrays.asList("1201", "user_1", "resource_3"),
            Arrays.asList("1202", "user_1", "resource_3")
        );

        List<List<String>> logs3 = Arrays.asList(
            Arrays.asList("300", "user_10", "resource_5")
        );

        printSolution(getSortedAccessTimes(logs1));
        List<String> res = getMostAccessedResource(logs1);
        System.out.println(res.get(0) + " : " + res.get(1));

        // Uncomment to test other logs
        // printSolution(getSortedAccessTimes(logs2));
        // printSolution(getSortedAccessTimes(logs3));
    }
}

File: coding/karat-log-access/sol.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <bits/stdc++.h>

using namespace std;

unordered_map<string, vector<int>> getSortedAccessTimes(vector<vector<string>> logs) {
  unordered_map<string, vector<int>> accesTimes;

  for(int i = 0; i<logs.size(); i++) {
    string user = logs[i][1];
    int num = stoi(logs[i][0]);

    if(accesTimes.find(logs[i][1]) != accesTimes.end()) {
      if(accesTimes[user][0] > num)
        accesTimes[user][0] = num;
      else if(accesTimes[user][1] < num)
        accesTimes[user][1] = num;
    }
    else {
      accesTimes[user] = {num,num};
    }
  }
  return accesTimes;
}

vector<string> getMostAccessedResource(vector<vector<string>> logs) {
  unordered_map<string, vector<int>> accesTimes;
  int max= 0;
  string resourceMax;


  for(int i = 0; i<logs.size(); i++) {
    string resource = logs[i][2];
    int num = stoi(logs[i][0]);

    while(accesTimes[resource].size() > 0 && accesTimes[resource][0] < num - 300) {
      accesTimes[resource].erase(accesTimes[resource].begin());
    }
    accesTimes[resource].push_back(num);

    if(accesTimes[resource].size() > max) {
      max = accesTimes[resource].size();
      resourceMax = resource;
    }
  }

  return {resourceMax, to_string(max)};
}

void printSolution(unordered_map<string, vector<int>> accessTimes) {
  for(auto itr = accessTimes.begin(); itr != accessTimes.end(); itr++) {
    cout << itr->first << " " << itr->second[0] << "," << itr->second[1] << "\n";
  }
}

int main() {
  vector<vector<string>> logs1 = {
      { "58523", "user_1", "resource_1" },
      { "62314", "user_2", "resource_2" },
      { "54001", "user_1", "resource_3" },
      { "200", "user_6", "resource_5" },
      { "215", "user_6", "resource_4" },
      { "54060", "user_2", "resource_3" },
      { "53760", "user_3", "resource_3" },
      { "58522", "user_22", "resource_1" },
      { "53651", "user_5", "resource_3" },
      { "2", "user_6", "resource_1" },
      { "100", "user_6", "resource_6" },
      { "400", "user_7", "resource_2" },
      { "100", "user_8", "resource_6" },
      { "54359", "user_1", "resource_3"},
  };

  vector<vector<string>> logs2 = {
      {"300", "user_1", "resource_3"},
      {"599", "user_1", "resource_3"},
      {"900", "user_1", "resource_3"},
      {"1199", "user_1", "resource_3"},
      {"1200", "user_1", "resource_3"},
      {"1201", "user_1", "resource_3"},
      {"1202", "user_1", "resource_3"},
  };

  vector<vector<string>> logs3 = {
      {"300", "user_10", "resource_5"},
  };

  printSolution(getSortedAccessTimes(logs1));
  vector<string> res = getMostAccessedResource(logs1);
  cout << res[0] << " : " << res[1];

//   printSolution(getSortedAccessTimes(logs2));
//   printSolution(getSortedAccessTimes(logs3));
  return 0;
}

// time - O(n)
// space for solution- O(n)
